### ISSUE TITLE
Is it a working contract?

### ISSUE THREAD

**Issue by naddika** [BryanMorgan/rent-share-solana-rust#1]
Is it a working contract? Can be used in a real environment?

**Comment by BryanMorgan** [comment]
Yes, this is a working contract. You can deploy it and test it yourself on Devnet or Mainnet. But I would consider this just a starter template for a real rental agreement since there are probably more features and validation you'd want to include besides just the basics here.

**Comment by naddika** [comment]
@BryanMorgan  I see. 

What those features and validation would be, for instance?

**Comment by BryanMorgan** [comment]
For example:

- Right now the contract is setup with a deposit and rental terms, but there is no instruction to refund the deposit. Should this happen immediately after the last payment or as a separate transaction, only allowed if the contract is in a complete state?
- There is an instruction to TerminateEarly, but under what conditions should that be allowed (or should it be removed?) and how can issue the instruction? Do you need both parties approval or just the owner (payee)?

**Comment by snowsky** [comment]
Hi @BryanMorgan, do you have an example for client side? Thanks :)

**Comment by BryanMorgan** [comment]
Right now I don't have a fully functional client-side example I can share. But I pulled out some of the key capabilities from a client example I am working on and shared it in this Gist. Hopefully that's enough to get you started with the 3 key steps:

- Create a rent share account
- Initialize the rental agreement with the terms (deposit, rent, duration, keys)
- Record a rental payment from a payer to a payee

It uses the `@solana/web3.js` library along with `BN` and `borsh` for big number and binary serialization support: 

https://gist.github.com/BryanMorgan/2da2ed2b1ce8371d12812cfc2db6e73b

**Comment by snowsky** [comment]
Thanks @BryanMorgan

**Comment by zigomi** [comment]
But that requires duplicating all those `AgreementStatus`, `RentDuration` and other things.... in JS?

**Comment by BryanMorgan** [comment]
No, you don't have to. You don't even have to use a JavaScript/web3.js client. You could write a client in Rust and re-use the same definitions.

But if you're building a web3-based client it probably makes it easier to keep track of the data you're sending in `TransactionInstruction`s if you model the contract schema data in JavaScript.
